haskell/private/actions/info.bzl

```diff
@@ -0,0 +1,181 @@
+"""Defines output groups that are consumed by tools such as 'hrepl'."""
+load("@bazel_skylib//lib:dicts.bzl", "dicts")
+load("@bazel_skylib//lib:paths.bzl", "paths")
+load(":providers.bzl", "all_package_ids", "get_ghci_extra_libs")
+load(
+    ":private/path_utils.bzl",
+    "get_lib_name",
+    "is_hs_library",
+)
+
+def write_proto_file(hs, output_name, proto_type, content):
+    """Write an encoded .proto file.
+
+    Writes a file with the text format encoding, and then runs "protoc"
+    to convert it to the wire encoding.
+
+    Using the wire encoding allows us use released versions of tools
+    with different versions of the Haskell rules (within reason).
+
+    Args:
+      hs: The current rule context.
+      output_name: The output filename.  The text-encoded file will be named
+        {output_name}.txt, and the encoded file will be named {file_name}.pb.
+      proto_type: The type of the proto (e.g. foo.Bar).  It must be defined
+        in rule_info.proto.
+      content: The contents of the text file, as a Bazel struct.
+
+    Returns:
+      A File containing the encoded proto message, named {file_name}.pb.
+    """
+    proto_txt = hs.actions.declare_file(output_name + ".txt")
+    proto_pb = hs.actions.declare_file(output_name + ".pb")
+    hs.actions.write(output = proto_txt, content = content.to_proto())
+
+    protoc = hs.toolchain.protoc
+    rule_info_protos = hs.toolchain.rule_info_proto[ProtoInfo].direct_sources
+
+    hs.actions.run_shell(
+        outputs = [proto_pb],
+        inputs = depset([proto_txt] + rule_info_protos),
+        tools = [protoc],
+        command =
+            "{protoc} {rule_info_proto} --encode {proto_type} < {proto_txt} > {proto_pb}"
+                .format(
+                protoc = protoc.path,
+                proto_type = proto_type,
+                proto_txt = proto_txt.path,
+                proto_pb = proto_pb.path,
+                rule_info_proto = " ".join([p.path for p in rule_info_protos]),
+            ),
+    )
+    return proto_pb
+
+# Copied from haskell/repl.bzl; not sure if there's a more general solution.
+_COMMON_REPL_OPTIONS = [
+    "-hide-all-packages",
+    "-package=base",
+    "-package=directory",
+]
```

Comment:
Are those needed here? In case of `haskell_repl` these are set because the GHCi script we use to load modules requires those packages. Does `hrepl` rely on these packages, too? If so, maybe it would be better to explicitly add those flags in `hrepl`, to keep the flags close to the code that relies on them.

Reply:
Yeah, that makes sense to me.  Done.

Reply:
To clarify, it's not that `hrepl` needs them, but `ghci` itself needs `base`:

```
$ stack exec ghci -- -hide-all-packages
<interactive>:1:6: error:
    Not in scope: ‘System.IO.hSetBuffering’
    No module named ‘System.IO’ is imported.
...
```

However I only added `base`, not `directory` which turned out not to be needed; I guess it's only used in your GHCi script.

Reply:
Thanks!